File: project/app/src/main/java/org/owntracks/android/ui/contacts/ContactsActivity.java
package org.owntracks.android.ui.contacts;

import android.content.Intent;
import android.os.Bundle;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.LinearLayoutManager;

import org.owntracks.android.R;
import org.owntracks.android.databinding.UiContactsBinding;
import org.owntracks.android.geocoding.GeocoderProvider;
import org.owntracks.android.model.FusedContact;
import org.owntracks.android.ui.base.BaseActivity;
import org.owntracks.android.ui.base.BaseAdapter;
import org.owntracks.android.ui.map.MapActivity;

import javax.inject.Inject;


public class ContactsActivity extends BaseActivity<UiContactsBinding, ContactsMvvm.ViewModel> implements ContactsMvvm.View, BaseAdapter.ClickListener<FusedContact> {
    private final ContactsAdapter contactsAdapter = new ContactsAdapter(this);

    @Inject
    GeocoderProvider geocoderProvider;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasEventBus(false);
        bindAndAttachContentView(R.layout.ui_contacts, savedInstanceState);
        setSupportToolbar(binding.toolbar);
        setDrawer(binding.toolbar);
        binding.getVm().getContacts().observe(this, contacts -> {
            contactsAdapter.setContactList(contacts.values());
            for (FusedContact contact : contacts.values()) {
                contact.getMessageLocation().removeObservers(binding.getLifecycleOwner());
                contact.getMessageLocation().observe(binding.getLifecycleOwner(), messageLocation -> geocoderProvider.resolve(messageLocation));
            }
        });
        binding.recyclerView.setLayoutManager(new LinearLayoutManager(this));
        binding.recyclerView.setAdapter(contactsAdapter);
    }

    @Override
    public void onClick(@NonNull FusedContact fusedContact, @NonNull View view, boolean longClick) {
        Bundle bundle = new Bundle();
        bundle.putString(MapActivity.BUNDLE_KEY_CONTACT_ID, fusedContact.getId());
        Intent intent = new Intent(this, MapActivity.class);
        intent.putExtra("_args", bundle);
        startActivity(intent);
    }
}

File: project/app/src/main/java/org/owntracks/android/ui/contacts/ContactsActivity.kt
package org.owntracks.android.ui.contacts

import android.content.Intent
import android.os.Bundle
import androidx.activity.compose.setContent
import androidx.activity.enableEdgeToEdge
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import dagger.hilt.android.AndroidEntryPoint
import javax.inject.Inject
import javax.inject.Named
import org.owntracks.android.data.repos.ContactsRepoChange
import org.owntracks.android.model.Contact
import org.owntracks.android.preferences.Preferences
import org.owntracks.android.support.ContactImageBindingAdapter
import org.owntracks.android.test.ThresholdIdlingResourceInterface
import org.owntracks.android.ui.map.MapActivity
import org.owntracks.android.ui.mixins.ServiceStarter
import org.owntracks.android.ui.navigation.Destination
import org.owntracks.android.ui.navigation.toActivityClass
import org.owntracks.android.ui.theme.OwnTracksTheme
import timber.log.Timber

@AndroidEntryPoint
class ContactsActivity :
    AppCompatActivity(),
    ServiceStarter by ServiceStarter.Impl() {

    @Inject
    @Named("contactsActivityIdlingResource")
    lateinit var contactsCountingIdlingResource: ThresholdIdlingResourceInterface

    @Inject
    lateinit var contactImageBindingAdapter: ContactImageBindingAdapter

    @Inject
    lateinit var preferences: Preferences

    private val viewModel: ContactsViewModel by viewModels()

    override fun onCreate(savedInstanceState: Bundle?) {
        enableEdgeToEdge()
        startService(this)
        super.onCreate(savedInstanceState)

        setContent {
            OwnTracksTheme(dynamicColor = preferences.dynamicColorsEnabled) {
                // Convert contacts map to a mutable state list sorted by timestamp
                val contactsList = remember {
                    mutableStateListOf<Contact>().apply {
                        addAll(viewModel.contacts.values.sortedByDescending { it.locationTimestamp })
                    }
                }

                // Observe contact changes
                androidx.compose.runtime.LaunchedEffect(Unit) {
                    viewModel.contactUpdatedEvent.collect { change ->
                        Timber.v("Received contactUpdatedEvent $change")
                        when (change) {
                            is ContactsRepoChange.ContactAdded -> {
                                contactsList.add(change.contact)
                                contactsList.sortByDescending { it.locationTimestamp }
                                viewModel.refreshGeocode(change.contact)
                            }
                            is ContactsRepoChange.ContactRemoved -> {
                                contactsList.removeAll { it.id == change.contact.id }
                            }
                            is ContactsRepoChange.ContactLocationUpdated -> {
                                val index = contactsList.indexOfFirst { it.id == change.contact.id }
                                if (index >= 0) {
                                    contactsList[index] = change.contact
                                    contactsList.sortByDescending { it.locationTimestamp }
                                }
                                viewModel.refreshGeocode(change.contact)
                            }
                            is ContactsRepoChange.ContactCardUpdated -> {
                                val index = contactsList.indexOfFirst { it.id == change.contact.id }
                                if (index >= 0) {
                                    contactsList[index] = change.contact
                                }
                            }
                            is ContactsRepoChange.AllCleared -> {
                                contactsList.clear()
                            }
                        }
                        contactsCountingIdlingResource.run { if (!isIdleNow) decrement() }
                    }
                }

                // Trigger geocode refresh on startup
                androidx.compose.runtime.LaunchedEffect(Unit) {
                    viewModel.contacts.values.forEach(viewModel::refreshGeocode)
                }

                ContactsScreen(
                    contacts = contactsList,
                    contactImageBindingAdapter = contactImageBindingAdapter,
                    onNavigate = { destination ->
                        navigateToDestination(destination)
                    },
                    onContactClick = { contact ->
                        startActivity(
                            Intent(this@ContactsActivity, MapActivity::class.java)
                                .putExtra(
                                    "_args",
                                    Bundle().apply {
                                        putString(MapActivity.BUNDLE_KEY_CONTACT_ID, contact.id)
                                    }
                                )
                        )
                    },
                    modifier = Modifier.fillMaxSize()
                )
            }
        }
    }

    private fun navigateToDestination(destination: Destination) {
        val activityClass = destination.toActivityClass() ?: return
        if (this.javaClass != activityClass) {
            startActivity(Intent(this, activityClass))
        }
    }
}
